Changelog 169 (#150)
intellij-rust/intellij-rust.github.io: _posts/2022-04-25-changelog-169.markdown

---
layout: post
title: "IntelliJ Rust Changelog #169"
date: 2022-04-25 13:00:00 +0300
---

## New Features

* [#8713] Update [ML completion﻿](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-code-completion.html#ml-completion) model. In addition to the data used before, the new model is now taking into account several new facts to:
- prefer expressions of type matching the expected type
- differentiate between inherent, trait and blanket implementations
- recognise `async`, `const` and `unsafe` contexts

* [#8699] Allow files generated by the `build.rs` buildscripts up to 8 megabytes in size. Previously is was necessary to adjust the limit manually using `idea.max.intellisense.filesize` property

* [#8697] Respect run/debug modes in [Run Anything](https://www.jetbrains.com/help/idea/running-anything.html) by Cargo commands

> * [#8675] Use `#[serde(rename)]` attribute for fields that do not use snake_case convention when converting from JSON (by [@Kobzol])

Reviewer: Since the feature is not enabled by default yet, let's drop this item

---
layout: post
title: "IntelliJ Rust Changelog #169"
date: 2022-04-25 13:00:00 +0300
---

## New Features

* [#6836] Initial support for [Run Targets](https://blog.jetbrains.com/idea/2021/01/run-targets-run-and-debug-your-app-in-the-desired-environment/).
  Run Targets allow you to build/run the applications you’re developing in Docker containers or on remote machines via SSH

  {% include gif-img.html path="/assets/posts/changelog-169/run-target" w="800px" %}

> * [#8699] Allow files generated by the `build.rs` buildscripts up to 8 megabytes in size.
>   Previously it was necessary to adjust the limit manually using `idea.max.intellisense.filesize` property.

Reviewer: ```suggestion
* [#8699] Process files generated by the [build scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html) up to 8 megabytes in size automatically.
  Previously such files were ignored by IDE during indexing and it was necessary to adjust the limit manually using `idea.max.intellisense.filesize` property to make IDE index them.
```

  Note, build script evaluation is disabled by default for now. To turn it on, enable `org.rust.cargo.evaluate.build.scripts` [experimental feature][experimental features]

> * [#8713] Update [ML completion] model.
>   In addition to the data used before, the new model is now taking into account several new facts to:
> 
>   * prefer expressions of type matching the expected type
> 
>   * differentiate between inherent, trait and blanket implementations
> 
>   * recognise `async`, `const` and `unsafe` contexts

Reviewer: ```suggestion
* [#8713] Update [ML completion] model.
  In addition to the data used before, the new model is now taking into account several new facts to:

  * prefer expressions of type matching the expected type

  * differentiate between inherent, trait and blanket implementations

  * recognise `async`, `const` and `unsafe` contexts
  
  Note that your source code is never being collected, only the information about your interactions with the code completion UI
```

> * [#8697] Respect run/debug modes in [Run Anything] by Cargo commands.
>   Now you can run a Cargo command under debugger using <kbd>Shift</kbd> + <kbd>Enter</kbd> in Run Anything popup.

Reviewer: ```suggestion
* [#8697] Respect run/debug modes in [Run Anything] by Cargo commands.
  Now you can run a Cargo command under debugger using <kbd>Shift</kbd> + <kbd>Enter</kbd> in Run Anything popup.
  It allows you to run you command in debug mode [from terminal](https://blog.jetbrains.com/idea/2020/07/run-ide-features-from-the-terminal/) as well
```

> * [#7998] Enable [Valgrind on WSL](https://www.jetbrains.com/help/clion/memory-profiling-with-valgrind.html#valgrind-wsl) toolchains

Reviewer: Let's be more precise here and use `Valgrind memcheck﻿`.
Also, I think it's worth mentioning that this integration works only in CLion

* [#7825] Add an option to use Rustfmt from a non-default toolchain and
  allow to configure addition arguments and environment variables for Rustfmt.

>   You can find these settings in `Settings | Languages & Frameworks | Rust | Rustfmt`

Reviewer: Let's move this item just after run targets

> * [#7410] Add inspection that finds unnecessarily qualified paths.

Reviewer: ```suggestion
* [#7410] Add `Unnecessarily qualified path` inspection that finds unnecessarily qualified paths.
```

>   It corresponds to the `unused_qualifications` lint (by [@Kobzol])

Reviewer: ```suggestion
  It corresponds to the `unused_qualifications` [lint](https://doc.rust-lang.org/rustc/lints/index.html) (by [@Kobzol])
```

## Fixes

* [#8739] Fix parsing of test events with large output in [Test View]

> * [#8742] Resolve unit structs in the context of a pattern.

Reviewer: ```suggestion
* [#8742] Resolve unit structs in the context of a pattern
```

* [#8726] Fix type inference of associated types that are referenced from a cast to a supertrait (by [@Arc-blroth])

* [#8703] Do not highlight [Documentation tests] if there are unbalanced code fences in a previous doc comment

* [#8738] Don't try to release editor of macro expansion view twice.

>   This fixes an exception [#8737] that occurs when closing a macro expansion popup.

Reviewer: ```suggestion
  This fixes an exception [#8737] that occurs when closing a macro expansion popup
```

* [#8706] Rename a URL scheme for macro expansion file system: `rust_macros` -> `rust-macros`.
  The previous name was incorrect because a URL scheme really can't contain an underscore sign `_`.

>   This fixes a possible exception during macro expansion.

Reviewer: ```suggestion
  This fixes a possible exception during macro expansion
```

* [#6836] Initial support for [Run Targets](https://blog.jetbrains.com/idea/2021/01/run-targets-run-and-debug-your-app-in-the-desired-environment/).
  Run Targets allow you to build/run the applications you’re developing in Docker containers, in WSL or on remote machines via SSH

  {% include gif-img.html path="/assets/posts/changelog-169/run-target" w="800px" %}

  Note that remote debugging using Run Targets is currently unavailable

* [#7825] Add an option to use Rustfmt from a non-default toolchain and
  allow to configure addition arguments and environment variables for Rustfmt.
  You can find these settings in `Settings | Languages & Frameworks | Rust | Rustfmt`

* [#8699] Process files generated by the [build scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html) up to 8 megabytes in size automatically.
  Previously such files were ignored by IDE during indexing and it was necessary to adjust the limit manually using `idea.max.intellisense.filesize` property to make IDE index them.
  Note, build script evaluation is disabled by default for now. To turn it on, enable `org.rust.cargo.evaluate.build.scripts` [experimental feature][experimental features]

> * [#8713] Update [ML completion] model.

Reviewer: ```suggestion
* [#8713] Update [ML completion] model to improve sorting of completion suggestions.
```

  In addition to the data used before, the new model is now taking into account several new facts to:

  * prefer expressions of type matching the expected type

  * differentiate between inherent, trait and blanket implementations

  * recognise `async`, `const` and `unsafe` contexts

  Note that your source code is never being collected, only the information about your interactions with the code completion UI

* [#8697] Respect run/debug modes in [Run Anything] by Cargo commands.
  Now you can run a Cargo command under debugger using <kbd>Shift</kbd> + <kbd>Enter</kbd> in Run Anything popup.
  It allows you to run your command in debug mode [from the terminal](https://blog.jetbrains.com/idea/2020/07/run-ide-features-from-the-terminal/) as well

* [#7998] Enable [Valgrind memcheck on WSL](https://www.jetbrains.com/help/clion/memory-profiling-with-valgrind.html#valgrind-wsl) toolchains.
  Note that `Valgrind Memcheck` integration is available in CLion only

> * [#7410] Add `Unnecessarily qualified path` inspection that finds unnecessarily qualified paths.

Reviewer: Maybe add a image/gif for the inspection?
